test/suite/wthooks.py

```diff
@@ -0,0 +1,256 @@
+            ret = replace_func(*args)
+        else:
+            ret = replace_func(self, *args)
+
+    for hook_func in notifies:
+        hook_func(ret, self, *args)
+    return ret
+
+# WiredTigerHookManager -
+# The hook manager class.  There is only one hook manager.  It is responsible for finding all the
+# HookCreators at the beginning of the run, and calling setup_hooks() for each one, to have it bind
+# hook functions to API functions.  The hook manager is initialized with a list of hook names. Each
+# name is expanded, for example, "demo" causes the hook manager to load hook_demo.py, and to call
+# the "initialize" global function in that file.  We expect "initialize" to return a list of objects
+# (hooks) derived from WiredTigerHook (class defined below).  Generally, "initialize" returns a
+# single object (setting up some number of "hook functions") but to allow flexibility for different
+# sorts of packaging, we allow any number of hooks to be returned.
+#
+# A hook can set up any number of "hook functions".  See hook_demo.py for a sample hook class.
+class WiredTigerHookManager(object):
+    def __init__(self, hooknames = []):
+        self.hooks = []
+        names_seen = []
+        for name in hooknames:
+            # The hooks are indicated as "somename=arg" or simply "somename".
+            # hook_somename.py will be imported, and initialized with the arg.
+            # Names must be unique, as we stash some info into extra fields
+            # on the connection/session/cursor, these are named using the
+            # unique name of the hook.
+            if '=' in name:
+                name,arg = name.split('=', 1)
+            else:
+                arg = None
+            if name in names_seen:
+                raise Exception(name + ': hook name cannot be used multiple times')
+            names_seen.append(name)
+
+            modname = 'hook_' + name
+            try:
+                imported = import_module(modname)
+                for hook in imported.initialize(arg):
+                    hook._initialize(name, self)
+                    self.hooks.append(hook)
+            except:
+                print('Cannot import hook: ' + name + ', check file ' + modname + '.py')
+                raise
+        for hook in self.hooks:
+            hook.setup_hooks()
+
+    def add_hook(self, clazz, method_name, hook_type, hook_func):
+        if not hasattr(clazz, method_name):
+            raise Exception('Cannot find method ' + method_name + ' on class ' + str(clazz))
+
+        # We need to set up some extra attributes on the Connection class.
+        # Given that the method name is XXXX, and class is Connection, here's what we're doing:
+        #    wiredtiger.Connection._XXXX_hooks = []
+        #    wiredtiger.Connection._XXXX_orig = wiredtiger.Connection.XXXX
+        #    wiredtiger.Connection.XXXX = lambda self, *args: hooked_function(self, 'XXXX', *args)
+        hooks_name = '_' + method_name + '_hooks'
```

Comment:
would like **hooks_name** to infer that its a list, maybe something like hooks_list here?

Reply:
Because of your previous comment, the hooks is no longer a list, it's a `WiredTigerHookInfo`, so I renamed this `hook_info_name`.